transformer_engine/jax/cpp_extensions.py

```diff
@@ -209,1824 +230,2350 @@ def custom_caller(name, args, opaque, has_side_effect, **kwargs):
 
-        out_types = [ir.RankedTensorType.get([ir_in_shape[1], ir_in_shape[0]], ir_out_dtype)]
-        operands = [inputs]
-        operand_shapes = [ir_in_shape]
+        out_shape = x_shape
+        hidden_size = reduce(operator.mul, w_shape)
+        batch_shape = out_shape[:-1]
+        batch_size = reduce(operator.mul, x_shape) // hidden_size
+
+        out_types = [
+            ir.RankedTensorType.get(out_shape, output_type),
+            ir.RankedTensorType.get(batch_shape, ir_mu_dtype),
+            ir.RankedTensorType.get(batch_shape, ir_rsigma_dtype),
+        ]
+        operands = [x, gamma, beta]
+        operand_shapes = [x_shape, w_shape, b_shape]
         args = CustomCallArgsWrapper(out_types, operands, operand_shapes)
 
-        assert len(ir_in_shape) == 2
-        opaque = transformer_engine_jax.pack_common_descriptor(ir_in_shape, dtype, dtype)
+        opaque = transformer_engine_jax.pack_norm_descriptor(
+            batch_size,
+            hidden_size,
+            jax_dtype_to_te_dtype(x_aval.dtype),
+            jax_dtype_to_te_dtype(gamma_aval.dtype),
+            zero_centered_gamma,
+            epsilon,
+        )
 
-        out = custom_caller(TransposePrimitive.name, args, opaque, False)
+        out = custom_caller(LayerNormFwdPrimitive.name, args, opaque, False)
 
-        return [out]
+        return out
 
+    @staticmethod
+    def impl(x, gamma, beta, zero_centered_gamma, epsilon):
+        """
+        to describe implementation
+        """
+        assert LayerNormFwdPrimitive.inner_primitive is not None
+        out, mu, rsigma = LayerNormFwdPrimitive.inner_primitive.bind(
+            x, gamma, beta, zero_centered_gamma=zero_centered_gamma, epsilon=epsilon)
+        return out, mu, rsigma
 
-_transpose_p = register_primitive(TransposePrimitive)
+    @staticmethod
+    def batcher(batched_args, batch_dims, *, zero_centered_gamma, epsilon):
+        """
+        to describe batch rules for vmap
+        """
+        assert LayerNormFwdPrimitive.outer_primitive is not None
+        x, gamma, beta = batched_args
+        x_bdim, _, _ = batch_dims
 
+        out_bdims = x_bdim, x_bdim, x_bdim
+        return LayerNormFwdPrimitive.outer_primitive.bind(x,
+                                                          gamma,
+                                                          beta,
+                                                          zero_centered_gamma=zero_centered_gamma,
+                                                          epsilon=epsilon), out_bdims
 
-def transpose(inputs: jnp.ndarray, dtype: TEDType) -> jnp.ndarray:
+    @staticmethod
+    def infer_sharding_from_operands(zero_centered_gamma, epsilon, mesh, arg_infos, result_infos):
+        del zero_centered_gamma, epsilon, result_infos
+        x_spec = get_padded_spec(arg_infos[0])
+        assert x_spec[-1] is None, \
+            f"Does not support to shard hidden dim in {LayerNormFwdPrimitive.name}"
```

Comment:
If you don't want to allow this, can you modify the message why and how to WAR this?
Something like:
`Does not support to shard hidden dim in {LayerNormFwdPrimitive.name} as this isn't efficient. If you still want this, can should be able to do it with jax.lax.with_sharding_constraint() to replicate that dimension.'

Reply:
Change to 
```
"Does not support to shard hidden dim in {LayerNormFwdPrimitive.name}, since it leads to incorrect results by missing necessary collective ops. To address, you could apply jax.lax.with_sharding_constraint to force no sharding on the hidden dim."
```

Reply:
What all native XLA operations does is to add the missing collective here by forcing None.
This is my preferred approach as it will make our custom operation behave like all native operations.

But I understood that you don't like this as it can easily cause a pattern that isn't fast.
So I wouldn't explain it as giving wrong value as we could just add the missing collective at the cost of perf.
So I would write something related to perf, not about valid values.

Did I miss something?

Reply:
I see. At this point of view, I agree to force no sharding in the hidden dim. I changed to "Force None in the hidden dim of LN's input PSpec". And adding corresponding warning when the given PSpec is not correct. Does this make more sense?